Use natural keys for foreign key relationships during serialization/deserialization

gherceg opened the pull request:
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->

## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->

I added what is hopefully a thorough description of this change in the comment for the dump method, but I'll rephrase that here and provide some additional proof for why this is necessary.

When loading domain data into a fresh environment from a file generated by the existing `dump_domain_data` command, we hit a ForeignKey violation when loading SQLUserData, which foreign keys to the Django auth.User object via the primary key.

The problem is that when we serialize our dumped data, we set `use_natural_primary_keys=True` to use natural keys for our sharded objects (the only place we defined `natural_key` methods in our code). However the auth.User defines a `natural_key` method, which ensures that the primary key is not included when serializing this object (docs [here](https://docs.djangoproject.com/en/3.2/topics/serialization/#natural-keys)). Since SQLUserData foreign keys to this object, and we have `use_natural_foreign_keys=False` in our serializer, the foreign key relationship is serialized using the auth.User's primary key, instead of its natural key.

Setting `use_natural_foreign_keys=True` ensures that the serialized SQLUserData will reference auth.User's natural key. Also have glossed over it, but auth.User also defines a `get_by_natural_key` method which is used when deserializing data, enabling SQLUserData to foreign key to the correct object at that point.

Here's an example of what a serialized SQLUserData object looks like before and after this change, looking most closely at the value for the "django_user" field:

Before
```
{"model": "users.sqluserdata", "pk": 1, "fields": {"domain": "gherceg", "user_id": "32ea2287870244ca8c448b1664be4f41", "django_user": 2, "modified_on": "2023-12-12T03:09:48.924945Z", "profile": null, "data": {}}}
```
After
```
{"model": "users.sqluserdata", "pk": 1, "fields": {"domain": "gherceg", "user_id": "32ea2287870244ca8c448b1664be4f41", "django_user": ["gherceg+admin@dimagi.com"], "modified_on": "2023-12-12T03:09:48.924945Z", "profile": null, "data": {}}}
```

The reason this came up recently is because SQLUserData is the first object we include in dumped data that foreign keys to auth.User. All other objects that foreign key to auth.User are excluded from the dump.

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
This code is run manually by developers and does not have any direct user facing impact. 

I confirmed the following:
- this does not impact the output of sharded objects that foreign key using form and case ids, since those foreign keys do not rely on the default model's primary key.
- verified that `natural_key` is only used on sharded objects in our code
- verified that `get_by_natural_key` is not used anywhere in our codej

The worst case is that this change corrupts the load_domain_data command, which is already broken because of this issue so 🤷🏻.
### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->
This area doesn't have the greatest test coverage. It is a bit tricky to ensure that _all_ models are dumped and loaded properly, but I intend to spend time improving the test suite here. When debugging this specific issue, I used some simple unit tests, but those didn't feel worth committing as they really just verify django serialization behavior as it is documented.
 
### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
N/A

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

gherceg committed:
Set use_natural_foreign_keys to True

dannyroberts reviewed (commented):
Is there any equivalent change that needs to be made to the load code, to make it know to interpret these as the "natural" key?

(My two comments are not blocking—I agree that this won't _break_ anything that currently works)

gherceg commented on corehq/apps/dump_reload/sql/dump.py:
```diff
@@ -237,9 +251,10 @@ def dump(self, output_stream):
         JsonLinesSerializer().serialize(
             objects,
-            use_natural_foreign_keys=False,
+            use_natural_foreign_keys=True,
```
Yeah good question. I was actually going to ask @snopoke if he had any ideas but it was also introduced a long time ago so wasn't sure how much context he would have. Looks like it was introduced [here](https://github.com/dimagi/commcare-hq/commit/506d966048f066de259c6e985a0f744ecdf9183e) if that helps jog anyone's memory.

The reason I didn't mention it is because my understanding is that this is a pretty safe setting. It _only_ impacts models that have `natural_key` defined, **and** are part of foreign key relationships that use the primary key. Our form and case models are the only place where both `natural_key` is defined and foreign key relationships are used. All relationships rely on form_id and case_id which are not the primary key for XFormInstance and CommCareCase respectively.

Even if they did rely on the primary key, we would want this setting enabled for the same reason we enable `use_natural_primary_key`, so that the natural key would be used to identify the object in the foreign key relationship. We would have to also define `get_by_natural_key` though to ensure the models are loaded properly. I can explain more about that method in my response to your other question.

gherceg replied:
Though looking at [django source code](https://github.com/django/django/blob/stable/3.2.x/django/core/serializers/python.py#L51-L60), I'm a bit confused about why the foreign keys for case and form objects aren't impacted by this. The code seems to suggest if `use_natural_foreign_keys` is defined and a `natural_key` method is defined, use the natural key instead of the value set for that field.

I think this is actually what is happening, and the "problem" is that our `natural_key` methods return a string (in the form of the case_id or form_id), as opposed to a tuple as the [docs specify](https://docs.djangoproject.com/en/3.2/topics/serialization/#serialization-of-natural-keys).

gherceg commented:
Danny asked a couple of important questions that have led to some more discoveries.

> Is there any equivalent change that needs to be made to the load code, to make it know to interpret these as the "natural" key?

Through testing this and the behavior I observed, my understanding was that this change would only result in a different output for the SQLUserData model, and not for Form and Case related models. My unjustified rationale was that the foreign key fields for Form and Case related models didn't reference the primary key directly, and that this somehow impacted how they were serialized, whereas the SQLUserData foreign key field to auth.User depended on the primary key.

However this is not the case. I think this behavior is due to how our `natural_key` methods for form and case related models are defined. As I mentioned in the thread to Danny's other question, the [django docs](https://docs.djangoproject.com/en/3.2/topics/serialization/#serialization-of-natural-keys) explicitly mention: 
> [The natural_key] method should always return a natural key tuple – in this example, (first name, last name).

So the `natural_key` method is being used for foreign key fields to CommCareCase and XFormInstance based on the Django serialization [code](https://github.com/django/django/blob/stable/3.2.x/django/core/serializers/python.py#L51-L60), but since CommCareCase and XFormInstance object managers do not define a `get_by_natural_key` method _and_ the `natural_key` method does not return a tuple, Django's [deserialization code](https://github.com/django/django/blob/stable/3.2.x/django/core/serializers/base.py#L301-L322) does not identify this field's value as a natural key.

All that to say, I don't foresee any immediate issues with this change, but we are not aligned with Django's recommended approach to using natural keys which could lead to unknown issues in future Django upgrades. Perhaps it is worth aligning with Django's recommendation just to be safe.

gherceg commented:
Actually, it looks like we do return a tuple for LedgerTransaction, which means as it stands right now, data generated with this change would not be successfully imported. I'm leaning towards doing what I said above, which will entail returning tuples for all `natural_key` methods and defining `get_by_natural_key` methods for each model that uses natural keys.

gherceg committed:
Add get_by_natural_key to default object managers

This updated any natural_key methods to ensure tuples were returned
which ensures the serialized keys will be usable in get_by_natural_key.

gherceg committed:
Fix failing JSONFieldSerialization test

gherceg committed:
Do not return tuples in natural_key for CommCareCase and XFormInstance

This goes against Django's recommendation and has the potential to
change behavior in future Django upgrades. However it is worth doing to
both accommodate our custom SQL load code, and to avoid needing to do DB
lookups for cases and forms for every object that foreign keys to
either.

gherceg committed:
Add tests to verify behavior of dumping and loading foreign key fields

gherceg committed:
Add test for SQLUserData in test_sql_dump_load

gherceg commented:
More updates! I ended up reverting back to returning the case_id and form_id for CommCareCase and XFormInstance instead of a tuple in 0d29b4e7f738ca962ad1df6c607541c178d900f8. After looking closer at Django's [deserializer code for foreign keys](https://github.com/django/django/blob/stable/3.2.x/django/core/serializers/base.py#L301-L322) with Daniel, the implications of returning a tuple as the natural key for CommCareCase and XFormInstance became more clear. Given that code, we would be incurring an additional DB lookup for a case or form when loading in large tables like CaseTransaction. There doesn't seem to be any reason we _need_ to return tuples at the current time, but as I mentioned in comments in that change, Django upgrades could potentially break this in the future.

So the remaining piece was to ensure foreign key fields for both iterable natural keys and non-iterable natural keys are loaded correctly. I added tests in 65cdf6cd3e6a65af50089902ca7e326964e71ceb testing both of those paths which confirm data is loaded as expected. I figured I could use a full "end to end" test to ensure SQLUserData is read from our DB, serialized, deserialized, and loaded as expected so added that in ce3cb54d69979a976c5028345abec402fafc3419.

gherceg committed:
Add tests to cover models withou natural_key defined

Includes some general cleanup as well

gherceg commented on corehq/form_processor/models/cases.py:
```diff
@@ -342,6 +342,13 @@ def __init__(self, *args, **kwargs):
 
     def natural_key(self):
+        """
+        Django recommends always returning a tuple in natural_key methods:
```
I know we read the same sentence together on a call, but since there isn't an enforcement of this at the code level, I was intentionally shying away from saying "requires", but I'm open to being convinced why "recommends always returning" isn't sufficient 😁

millerdev replied:
The docs say

> That method should always return a natural key tuple...

To me that sounds more like a requirement than a recommendation. The fact that we have gotten away with returning something other than a tuple is somewhere between convenient coincidence and wildly lucky.

gherceg replied:
52a70116718b0a05dd244e1e255419dfc0f6eb77

gherceg commented on corehq/apps/dump_reload/sql/dump.py:
```diff
@@ -229,6 +229,20 @@ class SqlDataDumper(DataDumper):
+        When serializing data using JsonLinesSerializer().serialize(...), the additional parameters are set for
+        the following reasons:
+
+        use_natural_primary_keys is necessary for sharded models to ensure the primary key, which will not be
```
Natural primary keys actually apply to _all_ sharded models, since all sharded models have the same issue with primary keys not being unique across shards. Does that explain this comment or is there still confusion here?

minhaminha replied:
No confusion - you can actually just ignore this suggestion

gherceg commented on corehq/apps/dump_reload/sql/dump.py:
```diff
@@ -229,6 +229,20 @@ class SqlDataDumper(DataDumper):
+        use_natural_primary_keys is necessary for sharded models to ensure the primary key, which will not be
+        unique across shards, is not used.
+
+        use_natural_foreign_keys is necessary for fk relationships where the primary key is used. For instance,
```
Ah sure. I'll include the bit about the `natural_key` method needing to be defined. Is the example with SQLUserData/auth.User still helpful?

minhaminha replied:
Yea I think keeping the example is good.

gherceg replied:
52a70116718b0a05dd244e1e255419dfc0f6eb77

gherceg committed:
Improve clarity of comments

gherceg commented on corehq/form_processor/models/cases.py:
```diff
@@ -342,6 +342,13 @@ def __init__(self, *args, **kwargs):
+        Django recommends always returning a tuple in natural_key methods:
+        https://docs.djangoproject.com/en/3.2/topics/serialization/#serialization-of-natural-keys
+        We intentionally do not follow this to optimize corehq.apps.dump_reload.sql.load.SqlDataLoader when other
+        models foreign key to CommCareCase or XFormInstance. This means our loader code is subject to break in
```
52a70116718b0a05dd244e1e255419dfc0f6eb77
